submissions/models.py

```diff
@@ -557,3 +560,139 @@ class Meta:
 
     creator = AnonymizedUserIDField()
     reason = models.TextField()
+
+
+class ExternalGraderDetailManager(models.Manager):
+    """
+    Manager for handling queue-related operations on Submissions.
+    """
+
+    def get_queue_length(self, queue_name):
+        """Count pending submissions in a specific queue"""
+        return self.time_filter().filter(
+            queue_name=queue_name,
+            status='pending'
+        ).count()
+
+    def get_next_submission(self, queue_name):
+        """Safely retrieve the next available submission for processing"""
+        return self.time_filter().filter(
+            queue_name=queue_name,
+            status='pending'
+        ).select_related('submission').order_by(
+            'created_at'
+        ).select_for_update().first()
+
+    def time_filter(self):
+        """
+        Filter submissions based on processing delay window.
+        Ensures we don't process submissions that were recently updated.
+        """
+        processing_window = now() - timedelta(
+            minutes=getattr(settings, 'SUBMISSION_PROCESSING_DELAY', 60)
+        )
+        return self.filter(status_time__lte=processing_window)
+
+
+class ExternalGraderDetail(models.Model):
+    """
+    Tracks queue processing information for a Submission.
+    """
+
+    STATUS_CHOICES = [
+        ('pending', 'Pending'),
+        ('pulled', 'Pulled'),
+        ('retired', 'Retired'),
+        ('failed', 'Failed'),
+    ]
+
+    VALID_TRANSITIONS = {
+        'pending': ['pulled', 'failed'],
+        'pulled': ['retired', 'failed'],
+        'failed': ['pending'],
+        'retired': []
+    }
+    submission = models.OneToOneField(
+        'submissions.Submission',
+        on_delete=models.CASCADE,
+        related_name='queue_record'
```

Comment:
Why is this different from the model name, i.e. why not make the model `QueueRecord`?